tchannel/container/heap.py

```diff
@@ -0,0 +1,90 @@
+from __future__ import absolute_import
+from __future__ import division
+
+import math
+import six
+
+
+class HeapOperation(object):
+    def cmp(self, i, j):
+        """Compare the items in position i and j of the heap.
+
+        :param i: the first item's position of the heap list
+        :param j: the second item's position of the heap list
+        :return true if items[i] < items[j], otherwise false.
+        """
+        raise NotImplementedError()
+
+    def push(self, x):
+        """Push a new item into heap"""
+        raise NotImplementedError()
+
+    def pop(self):
+        """Pop an item from the heap"""
+        raise NotImplementedError()
+
+    def swap(self, i, j):
+        """swap items between position i and j of the heap"""
+        raise NotImplementedError()
+
+
+def init(h):
+    # heapify
+    n = len(h)
+    for i in six.moves.range(int(math.floor(n/2)) - 1, -1):
```

Comment:
Because GitHub hid the comment, just a reminder:

     range(n, -1) == []` for n >= 0

You'll need comprehensive separate tests for the heap implementation.